### DISCUSSION TITLE
Choice of technology

### DISCUSSION THREAD

**Discussion by escamoteur** [DevsHelpDevs/DevsHelpDevs#5] [Q&A]
This probably is the question most of youo will like most to discuss about.

As I don't know that much about frontend development I need your opinions on that as well on which would be the best approach for such a platform. Also we have to agree on which sort of backend we want to use.

**Comment by sameeerkashyap** [comment]
I would suggest React + firebase for the most stable build and reliable framework right now, but I have very little experience in React myself and we might have to reach out to the react community. 

Although if we can wait a bit longer, I'm anticipating Flutter might bring Flutter Web to stable on the 3/3/21 event.
 
Then we could go for flutter, as all of us can work on it flexibly.
 And I believe that the flutter team will keep pushing to more stable throughout this year.

**Comment by escamoteur** [comment]
I guess we need here the input of web experts. I know many love Vue over React.

For the Backend I propose https://nhost.io/
I already contacted them if they would be willing to host us for free. As it's a young Scandinavian company I hope the might join us.

**Reply by sameeerkashyap** [reply]
So you suggest serverless functions using this ?

**Reply by escamoteur** [reply]
Not necessarily. With the GraphQL interface I don't know how much custom server logic we will need. For authentication they also have an API

**Reply by sameeerkashyap** [reply]
Yes, GraphQL would make things very simple, especially CRUD. Sounds good

**Reply by kevin-sakemaer** [reply]
Do you really need a backend ?
What I mean by that is firebase can do all of this.

If the goal is to have a custom backend then what can it do ? I think we need to have a global view of what will be possible inside the app. And then we can make enlightened choice.


Finally with @jaumard we have some skill in using dart for this purpose 😆

**Reply by escamoteur** [reply]
@Kleak Firebase isn't free and we got free hosting of database, graphql and auth be nhost.io.
If we need some custom functionality they even offer their version of clod functions

**Reply by kevin-sakemaer** [reply]
ok.
You are right it's not free that depends on the traffic you got because there is a free tier. 
And from their website their free tier is lot more restrictive than the one on firebase.

**Reply by escamoteur** [reply]
@Kleak you may not have understand me, they sponsor us the backend not the free tier :-)

**Reply by escamoteur** [reply]
They also will have a Flutter SDK the next weeks

**Reply by kevin-sakemaer** [reply]
Great. 
But as a Dart fan, that would just be really great to push it everywhere and obviously here also :p

**Reply by escamoteur** [reply]
I pushed them to create a Dart SDK :-)

**Reply by escamoteur** [reply]
Also you don't get that level of direct personal support with firebase :-)

**Reply by kevin-sakemaer** [reply]
> I pushed them to create a Dart SDK :-)

Not really the same as having a full backend written in dart but a good step.

> Also you don't get that level of direct personal support with firebase :-)

Probably

**Reply by escamoteur** [reply]
@Kleak if you want to join us https://join.slack.com/t/devshelpdevs/shared_invite/zt-nt662pjn-RuaamoGvjSbHhEqtfRXVIg

**Comment by escamoteur** [comment]
If we go with NHOST, we just got free hosting donated by them
<img width="957" alt="grafik" src="https://user-images.githubusercontent.com/16635729/106456100-5d18ba00-648d-11eb-8684-931bd8b530a1.png">

**Reply by sameeerkashyap** [reply]
Best opportunity to go with them. 
Sounds like a plan!!

**Reply by escamoteur** [reply]
Unless noone else has some strong argument against them

**Reply by StephenLadoye** [reply]
Yea, sounds like best option for now

**Comment by escamoteur** [comment]
While writing a preliminary user specification I change my mind a bit. I think a Web App for potential donors is probably the best. for people applying a Flutter App might be the better solution as it makes it way easier to upload pictures and gives us additional possibilities to prevent fraud.

**Reply by sameeerkashyap** [reply]
So are you suggesting building for web + mobile using flutter?

**Reply by escamoteur** [reply]
Would be my proposal at the moment but I'm open for other ideas

**Reply by trojanpauro** [reply]
i suggest that each platform  offers almost the same functions

**Reply by escamoteur** [reply]
This can be the goal in the end, but to get fast to results I would limit the functionality to the different platforms. If we choose to implement parts as apps and parts on web

**Reply by TheWestDevop** [reply]
Alright that's a great idea, we can start with the web and with time move to app

**Reply by jaumard** [reply]
Flutter FTW :)

**Reply by escamoteur** [reply]
We use Flutter for the two Apps that we will develop too :-)

**Reply by escamoteur** [reply]
@jaumard if you want to join use, here is our slack https://join.slack.com/t/devshelpdevs/shared_invite/zt-nt662pjn-RuaamoGvjSbHhEqtfRXVIg

**Comment by TheWestDevop** [comment]
> This probably is the question most of youo will like most to discuss about.
> As I don't know that much about frontend development I need your opinions on that as well on which would be the best approach for such a platform. Also we have to agree on which sort of backend we want to use.

We can actually use any front end tools but we should consider what most people are conversant with for easy maintenance

**Comment by biggaji** [comment]
For the front end, am thinking we should render the app from the server using handlebars or any templating engine.
Render from the server is fast.
Graphql is good for the api calls, then I suggest we used postgres for the database.
What do you think?

**Reply by escamoteur** [reply]
Interesting approach. so no react or vue or sventje? We probably use https://nhost.io/ as backend, they give us free hosting. they use prostgres

**Reply by biggaji** [reply]
Wow! That is great then.

**Reply by 0xdanny** [reply]
> For the front end, am thinking we should render the app from the server using handlebars or any templating engine.

A templating engine seems like neutral ground, that's if we aren't going to go with Flutter. But how flexible is this? I have very little web dev experience but i know sites built that way are not always very scalable.

> Render from the server is fast.

Agreed. But we shouldn't do that with templating engines, see [When not to use Handlebars](https://handlebarsjs.com/installation/when-to-use-handlebars.html). If the web devs here think SSR is right for this, i think it'll have to be done with Vue/React (Angular anybody? 😄 )

**Reply by escamoteur** [reply]
I don't think Flutter is an Option for the Website. For a pure web experience Web Technology will probably stay the best way to go. https://svelte.dev/ could be interesting, at least I hear a lot of good stuff about it.

**Reply by wilsonowilson** [reply]
I agree. Plus, in countries with slower and more expensive internet, lower page sizes are a must. Flutter is just too large to ship atm.

**Reply by 0xdanny** [reply]
> A templating engine seems like neutral ground, that's if we aren't going to go with Flutter. But how flexible is this? I have very little web dev experience but i know sites built that way are not always very scalable.

I actually meant if we aren't going to go with a Flutter mobile app, not website 😅 

I totally agree with y'all though. Svelte is very promising but do we have web developers here (not necessarily Svelte cause i believe it has a shallow learning curve). I'm very eager to see development on this get started as i just got a whole lot of free time.

**Reply by escamoteur** [reply]
I still think the applicant and the Staff App could be a Flutter app.

**Reply by 0xdanny** [reply]
> I still think the applicant and the Staff App could be a Flutter app.

> While writing a preliminary user specification I change my mind a bit. I think a Web App for potential donors is probably the best. for people applying a Flutter App might be the better solution as it makes it way easier to upload pictures and gives us additional possibilities to prevent fraud.

I agree. This is the best as you've said earlier.

**Reply by 805702** [reply]
Sites built that way are not scalable at all.

**Reply by escamoteur** [reply]
@805702 built in which way?

**Reply by 805702** [reply]
I actually don't agree with the SSR idea. This is something that seems easier said than done. the amount of expertise needed for this to be implemented is not small. and also, SSR is very good but does not stand for dashboard applications or applications that need to be shared on social media.

**Reply by 805702** [reply]
So based on this I propose we make a React.js app for the dashboard and if not React.js for the applicant side, we use Svelte.js that is well adapted for SEO and permits easy sharing.

**Reply by 805702** [reply]
IMHO

**Reply by escamoteur** [reply]
@805702 which technology would be SSR? Sorry I don't know what it stands for.

**Reply by 805702** [reply]
https://github.com/DevsHelpDevs/DevsHelpDevs/discussions/5?sort=new#discussioncomment-378504

sites built with server side rendering.

**Reply by 805702** [reply]
https://github.com/DevsHelpDevs/DevsHelpDevs/discussions/5?sort=new#discussioncomment-379131

SSR stands for Server Side Rendering.
for example using handlebars seems easy but loading a whole application based on that becomes a problem especially when it is a dashboard app.

and this platform is going to be a dashboard application.

**Reply by 0xdanny** [reply]
i agree with @805702. That's what i meant [earlier](https://github.com/DevsHelpDevs/DevsHelpDevs/discussions/5#discussioncomment-375556) that sites built with a templating engine (like handlebars) aren't always scalable, especially with SSR.

Svelte would be okay though as @escamoteur said. I think we should just go with the usual stuff, Svelte/Vue and Nhost.io for the backend.
